File: LeetCode/Study Plan/Practice Problems/Medium/As Far From Land As Possible/Go/asFarFromLandAsPossible.go
package main

import "fmt"

var dir = [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}

func maxDistance(grid [][]int) int {
	size := len(grid)
	distance := -1
	q := [][2]int{}
	visited := make([][]int, size)
	for i := 0; i < size; i++ {
		visited[i] = make([]int, size)
	}

	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			visited[i][j] = grid[i][j]
			if grid[i][j] == 1 {
				q = append(q, [2]int{i, j})
			}
		}
	}

	for len(q) > 0 {
		qLen := len(q)
		for k := 0; k < qLen; k++ {
			curr := q[0]
			q = q[1:]
			for i := 0; i < 4; i++ {
				x, y := curr[0]+dir[i][0], curr[1]+dir[i][1]
				if x >= 0 && y >= 0 && x < size && y < size && visited[x][y] == 0 {
					visited[x][y] = 1
					q = append(q, [2]int{x, y})
				}
			}
		}
		distance++
	}

	if distance == 0 {
		return -1
	}
	return distance
}

func main() {
	fmt.Println(maxDistance([][]int{
		{1, 0, 1}, {0, 0, 0}, {1, 0, 1},
	}))
}

File: LeetCode/Medium/As Far From Land As Possible/Java/asFarFromLandAsPossible.java
class Solution {
    int[][] dir = new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    public int maxDistance(int[][] grid) {
        int[][] visited = new int[grid.length][grid.length];
        ArrayDeque<Pair<Integer, Integer>> queue = new ArrayDeque<>();
        int size = grid.length;
        int distance = -1;

        for (int i=0; i<size; i++) {
            for (int j=0; j<size; j++) {
                visited[i][j] = grid[i][j];
                if (grid[i][j] == 1) {
                    queue.offer(new Pair(i, j));
                }
            }
        }

        while(!queue.isEmpty()) {
            int qLen = queue.size();

            while(qLen-- > 0) {
                Pair<Integer, Integer> curr = queue.poll();
                for (int[] d : dir) {
                    int x = d[0] + curr.getKey();
                    int y = d[1] + curr.getValue();
                    if (x >= 0 && y >= 0 && x < size && y < size && visited[x][y] == 0) {
                        queue.offer(new Pair(x, y));
                        visited[x][y] = 1;
                    }
                }
            }
            distance++;
        }

        return distance == 0 ? -1 : distance;
    }
}
